### DISCUSSION TITLE
Connection problems after Server update

### DISCUSSION THREAD

**Discussion by januweness** [BlazejosP/huawei-sun2000-API-CLI#14] [Q&A] [answered]
Hi, after the server update, the script no longer works. I have been communicating with Huawei support for some time, and they already discovered on syntax problem in line 556 of functions.sh:

logowanie=$(echo '{userName: "'$userName'", systemCode: "'$systemCode'"}'| http --print=hb --follow --timeout 3600 POST https://eu5.fusionsolar.huawei.com/thirdData/login  Content-Type:'application/json'  Cookie:'Cookie_1=value; web-auth=true;')
=>
logowanie=$(echo '{"userName": "'$userName'","systemCode": "'$systemCode'"}'| http --print=hb --follow --timeout 3600 POST https://eu5.fusionsolar.huawei.com/thirdData/login  Content-Type:'application/json'  Cookie:'Cookie_1=value; web-auth=true;')

But that has also not solved the problem. 

The screenoutput is:
Please wait connecting!
Connection and login to API with given Username&Password
 set-cookie: XSRF-TOKEN=x-rwqqo7vwmk7y9d07bvfy5d7tdiurnulf48ka6m07s55h3srvc908s7jzimg7obfu87djca5dc4k6bxvs2mjs4art2r88mls97zal2p1ihdlilcbxlilj08uldhrx9f3u;Path= xsrf-token: x-rwqqo7vwmk7y9d07bvfy5d7tdiurnulf48ka6m07s55h3srvc908s7jzimg7obfu8 {"data":null,"success":true,"failCode":0,"params":{},"message":null}

So, the success flag is set to positive, but I still get:
Problems with conection to Huawei Server

Network Error :(

So it seems that the ouput of the above command has changed with the server update and can no longer be properly interpreted?

**Comment by xvanhien** [comment]
Same issue here:
XSRF-TOKEN:
JSESSIONID:


Problems with conection to Huawei Server

Network Error :(
Returned data:

**Comment by januweness** [comment]
Apparently, the output syntax has changed, and the separators like
IFS=';'
and the array indices like
xsrf_token=${array[1]}
need to be redone. I just see in the issues, that this has been understood and planned to be fixed by end of May.

Thanks so much to the developers for the great work!

**Comment by BlazejosP** [comment]
Hello I'm happy that you are interested in this issue
, Software stop working because this line 561 in functions.sh and userName and  systemCode now needs this "
So congratulations for @januweness  he solved the problem first correctly but things are more complex to made software working again. As @xvanhien found congratulations for that. Problem is with extract xsrf-token from answer which is not extracted correctly now after Huawei upgrade.

I upload you partially working code to repository which can login but still can't extract xsrf-token functions.sh my changes are now in lines between 538 to 572 . I'm working on this now but rebuilding  function login_to_API may take me some time so if someone of you can upload working solution faster I will be really happy to accept your contribution. I send you both invitation for collaboration so you can create pull request. I also hope that all will work correctly after rebuilding only this one function but  who now hope there will be not much necessary changes.

**Comment by januweness** [comment]
Hi, sure, I will be happy to help. Note, though, that the invitation
leads to 404.

cheers,
Jan-Uwe

On 20/5/21 23:35, BlazejosP wrote:
>
> Hello I'm happy that you are interested in this issue, Software stop
> working because this line 561 in functions.sh and userName and
> systemCode now needs this "
> So congratulations for @januweness <https://github.com/januweness> he
> solved the problem first correctly but things are more complex to made
> software working again. As @xvanhien <https://github.com/xvanhien>
> also congratulations found problem is with extract xsrf-token from
> answer which is not extracted correctly.
>
> I upload you partially working code which can login but still can't
> extract xsrf-token. I'm working on this now but rebuilding function
> login_to_API may take me some time so if someone of you can upload
> working solution faster I will be really happy to accept your contribution
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlazejosP/huawei-sun2000-API-CLI/discussions/14#discussioncomment-765622>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTM3ACYOTLZW7GJRTJQL63TOV6BLANCNFSM44ZLTGMA>.
>


-- 

P.S. This e-mail might reach you during non-working hours. This is a
matter of my personal work-life balance and travel schedule only. I do
not expect to receive a reply outside of your usual working hours.

Dr. Jan-Uwe Ness
Operations Scientist
Chair of ESAC Science Faculty
European Space Astronomy Centre
Camino Bajo del Castillo s/n
Urb. Villafranca del Castillo
28692 Villanueva de la Canada
Madrid, Spain
http://januweness.eu <http://januweness.eu>
ORCID 0000-0003-0440-7193


This message is intended only for the recipient(s) named above. It may contain proprietary information and/or
protected content. Any unauthorised disclosure, use, retention or dissemination is prohibited. If you have received
this e-mail in error, please notify the sender immediately. ESA applies appropriate organisational measures to protect
personal data, in case of data privacy queries, please contact the ESA Data Protection Officer ***@***.***).

**Reply by BlazejosP** [reply]
I have in panel that  invitation is waiting for your acceptance. Can you check again and check if you can made some changes in repository now?

**Reply by BlazejosP** [reply]
Now you are collaborator I noted this in settings

**Reply by BlazejosP** [reply]
So I will work on this tomorrow morning  now go to sleep  but this is answer what we have with this new uploaded version

```
./fusionsolarapp.sh
Please wait connecting!
Connection and login to API with given Username&Password

 set-cookie: XSRF-TOKEN=x-eos7dfg9bzs9fy8akbtj6ns7kbemfukaqo2ps67z5glg5g7us9lhaovuaopgurbxs99irs1eqnipqo4amrs9hgpd6ok98b8bnzeoqr2kinfuvtgbs75h2qerjxvwmqjw;Path= xsrf-token: x-eos7dfg9bzs9fy8akbtj6ns7kbemfukaqo2ps67z5glg5g7us9lhaovuaopgurbxs {"data":null,"success":true,"failCode":0,"params":{},"message":null}

Problems with connection to Huawei Server

Network Error :(
Returned data: 

```

As you see "success":true  but still there is Network Error :( now is necessary to extract this xsrf-token with bash hope that is only change now necessary to make everything working

**Comment by januweness** [comment]
Sorry, I forgot to sign in. Now I have accepted and can edit.

 I will develop offline first for testing let you know when I get the
token extraction to work. At least that one should be straight forward,
just tedious, but I am happy to work on that.

Thanks,
Jan-Uwe

On 20/5/21 23:55, BlazejosP wrote:
>
> I have in panel that invitation is waiting for your acceptance. Can
> you check again and check if you can made some changes in repository now?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlazejosP/huawei-sun2000-API-CLI/discussions/14#discussioncomment-765692>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTM3AFJXNPTXHMAPW7X4ELTOWAOFANCNFSM44ZLTGMA>.
>


-- 

P.S. This e-mail might reach you during non-working hours. This is a
matter of my personal work-life balance and travel schedule only. I do
not expect to receive a reply outside of your usual working hours.

Dr. Jan-Uwe Ness
Operations Scientist
Chair of ESAC Science Faculty
European Space Astronomy Centre
Camino Bajo del Castillo s/n
Urb. Villafranca del Castillo
28692 Villanueva de la Canada
Madrid, Spain
http://januweness.eu <http://januweness.eu>
ORCID 0000-0003-0440-7193


This message is intended only for the recipient(s) named above. It may contain proprietary information and/or
protected content. Any unauthorised disclosure, use, retention or dissemination is prohibited. If you have received
this e-mail in error, please notify the sender immediately. ESA applies appropriate organisational measures to protect
personal data, in case of data privacy queries, please contact the ESA Data Protection Officer ***@***.***).

**Comment by BlazejosP** [comment]
Software is working again function login_to_API rebuild and working like before.  Test on your installations different functions hope will work also for you. Generally working also other functions but are some problems with handling dates so function which use actually time need some touch rebuid date handling.

**Comment by januweness** [comment]
Wow!!!! Thanks so much. Your are a crack ;-)

cheers,
Jan-Uwe

On 21/5/21 21:35, BlazejosP wrote:
>
> Software is working again function login_to_API rebuild and working
> like before.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlazejosP/huawei-sun2000-API-CLI/discussions/14#discussioncomment-769885>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTM3AGGM6TUK7WC5SUTQ23TO2YYZANCNFSM44ZLTGMA>.
>


-- 

P.S. This e-mail might reach you during non-working hours. This is a
matter of my personal work-life balance and travel schedule only. I do
not expect to receive a reply outside of your usual working hours.

Dr. Jan-Uwe Ness
Operations Scientist
Chair of ESAC Science Faculty
European Space Astronomy Centre
Camino Bajo del Castillo s/n
Urb. Villafranca del Castillo
28692 Villanueva de la Canada
Madrid, Spain
http://januweness.eu <http://januweness.eu>
ORCID 0000-0003-0440-7193


This message is intended only for the recipient(s) named above. It may contain proprietary information and/or
protected content. Any unauthorised disclosure, use, retention or dissemination is prohibited. If you have received
this e-mail in error, please notify the sender immediately. ESA applies appropriate organisational measures to protect
personal data, in case of data privacy queries, please contact the ESA Data Protection Officer ***@***.***).

**Comment by xvanhien** [comment, marked as answer]
I have same error, it isn't fixt... :-(


Please wait connecting!
Connection and login to API with given Username&Password


value {array2[18]} = true,"failCode"


value {array2[0]} = true
true
HTTP/1.1 200 OK
Server: product only
Date: Mon  24 May 2021 12:04:42 GMT
Content-Type: application/json;charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
set-cookie: XSRF-TOKEN=x-ul7y7y0b5gjzldpcirmno7bvc6s9c4mnrtde1d3xqp3wc7by1eem7sjvs8lg8646apljumvy1gc8de6mekc60506uk04eog8btjytgtibx9cs40auoc7c444rwby6p5j;Path=/                                                 ;Secure
x-trace-enable: false
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
xsrf-token: x-ul7y7y0b5gjzldpcirmno7bvc6s9c4mnrtde1d3xqp3wc7by1eem7sjvs8lg8646apljumvy1gc8de6mekc60506uk04eog8btjytgtibx9cs40auoc7c444rwby6p5j
x-frame-options: SAMEORIGIN
x-download-options: noopen
strict-transport-security: max-age=31536000; includeSubDomains
content-encoding: gzip

{"data":null "success":true "failCode":0 "params":{} "message":null}


value array3[40] = {"data":null,"success":true,"failCode":0,"params":{},"message":null}
value array3[39] =

Problems with conection to Huawei Server

Request success or failure flag:

Network Error :(
Returned data:


When I change: local sucesfully_login_true_or_false=$(echo ''${**array3[39]**}''  | jq '.success' )  to local sucesfully_login_true_or_false=$(echo ''${**array3[40]**}''  | jq '.success' )

then next step is wrong. What can I change?

**Reply by BlazejosP** [reply]
Don't know why you have still the same problem with login but can you upgrade  functions.sh and fusionsolarapp.sh to latest version which I modified today and rapport me how this versions are working hope together we found why is not working on your side.

**Reply by BlazejosP** [reply]
In my case  @BlazejosP and @januweness
this newest version as I understand is working and allows connection. Test this again with newest files if problem remain I'll try to help you in hunting bug which is reason why you can't login probably.

I add you as a contributor invitation is send to your account so you can add improvement yourself to repository if you will accept invitation.

**Reply by BlazejosP** [reply]
> I have same error, it isn't fixt... :-(

Hello @xvanhien I'm just curious how is going on on your side still have this problem with connection to API even with new versions of fusionsoalarapp.sh and  functions.sh ?

**Reply by xvanhien** [reply]
@BlazejosP I just download the versions of fusionsolarapp.sh and function.sh and have an error. And uncomment line 607+608+609 + 633.
Please wait connecting!
Connection and login to API with given Username&Password

Line 608: XSRF-TOKEN: x-ep7ukbbx5gg444o5nzuk9g7sakkbby05g45do7pclc2pk9vw0asb0405kbg4eptc1e46hco6unjygb482rvxeruoilvy6rtgvyob6p5ebumkg8k5tjnypeo75jjs1i9f

line 633: true

Problems with conection to Huawei Server

Network Error :(
Returned data:

**Reply by BlazejosP** [reply]
Looks like XSRF-TOKEN which is needed by all other functions is extracted correctly looking  on your rapport so this part of code is working correct. Something goes wrong with extracting from answer 
JOSN part and then from this JOSN part .success simple words true or false. So error must be somewhere here

functions.sh between lines 614 to 651
```
IFS=':'
local array2=( $logowanie_for_josn_extraction )
# showing diffrent values experimenting with postion in array
#echo ""
#echo ""
#echo "value = ${array2[18]}"

local logowanie_for_josn_extraction=${array2[18]}

IFS=','
local array2=( $logowanie_for_josn_extraction )
#echo ""
#echo ""
#echo "value = ${array2[0]}"

# finally true or false login data extracted
local sucesfully_login_true_or_false=$(echo ${array2[0]})


#echo $sucesfully_login_true_or_false

#echo $josn

# extracting JOSN from full response with headers
IFS=$' \r\n\r\n'
local array3=( $josn )
#show response from API in JOSN
#echo ""
#echo ""
#echo "value = ${array3[40]}"

if [[ $sucesfully_login_true_or_false =~ "true"  ]];
then
	local sucesfully_login_true_or_false=$(echo ''${array3[39]}''  | jq '.success' )
else
	local sucesfully_login_true_or_false=$(echo ''${array3[40]}''  | jq '.success' )
	
fi
```
Three posible solutions

1. For experiment try change values here #echo "value = ${array3[40]}" instead of 40 tray different numbers somewhere there is your word true probably or show whole array #echo "value = ${array3[@]}" 

for me word true was under position 39 and in case of error on position 40 word false

2. secound solution for experiment  try to comment this whole part of code in bash  in between
```
: <<'END_COMMENT'

END_COMMENT
```

and hardcoded 
```
local sucesfully_login_true_or_false="true"
```

I'm curious if in that case program will work with your account?

 If not my proposition is if you can give me a temporary access to your plant by some private channel like email (after that you can change password with Huawei helpdesk to new one) and I will see why your answer is different and what we can do to prevent this problem what you have in the future. Because my intuition now tell me that as for now you have sightly longer or shorter answer from function login_to_API and words true or false are in different positions in array but why?

**Reply by xvanhien** [reply]
Response solution 1:
Please wait connecting!
Connection and login to API with given Username&Password

XSRF-TOKEN: x-bu2mk9cbfs8b1jgbrxs7modinw5geompc7uonzg7c8qqhc3wnt843smngbapob9d1cfuo5042l4488dd2r5i1fupg9rubseodicb2p6qnss55dljmolhka86oaamirc8

true
value = HTTP/1.1 200 OK Server: product only Date: Thu, 27 May 2021 17:14:49 GMT Content-Type: application/json;charset=UTF-8 Transfer-Encoding: chunked Connection: keep-alive set-cookie: XSRF-TOKEN=x-bu2mk9cbfs8b1jgbrxs7modinw5geompc7uonzg7c8qqhc3wnt843smngbapob9d1cfuo5042l4488dd2r5i1fupg9rubseodicb2p6qnss55dljmolhka86oaamirc8;Path=/;Secure x-trace-enable: false x-xss-protection: 1; mode=block x-content-type-options: nosniff xsrf-token: x-bu2mk9cbfs8b1jgbrxs7modinw5geompc7uonzg7c8qqhc3wnt843smngbapob9d1cfuo5042l4488dd2r5i1fupg9rubseodicb2p6qnss55dljmolhka86oaamirc8 x-frame-options: SAMEORIGIN x-download-options: noopen strict-transport-security: max-age=31536000; includeSubDo


mains content-encoding: gzip  {"data":null,"success":true,"failCode":0,"params":{},"message":null}



Response solution 2, is also not working. 
When I have hardcoded **local josn_final=${array3[40]}** it works and comment line 647 till 697
[functions.txt](https://github.com/BlazejosP/huawei-sun2000-API-CLI/files/6555745/functions.txt)

**Reply by BlazejosP** [reply]
I made  this working based on your fuctions.txt I uncommented this what you commented and try made this working. I'm not certain but first check as is probably true in my case is in array position 39 and false position 40. For me now this working without problem.

As I understand in your case true meaby on position 40 and false on 41 check yourself. Meaby will be necessary to write some algorithm with will search through whole array for word true or false if different persons has slight different answers and positions of this words are fluid.

```
#show response from API in JOSN
#echo ""
#echo ""
#echo "value = ${array3[40]}"
#echo "value = ${array3[39]}"

#local sucesfully_login_true_or_false=$(echo ''${array3[39]}''  | jq '.success')

if [[ $sucesfully_login_true_or_false == "true"  ]];
then
	sucesfully_login_true_or_false=$(echo ''${array3[39]}''  | jq '.success' )
else
	sucesfully_login_true_or_false=$(echo ''${array3[40]}''  | jq '.success' )
	
fi


if [[ $sucesfully_login_true_or_false =~ "false"  ]];
	then		
			if [ ! -z "$DIALOG" ];
			then
				info_for_dialog_screen="Login to server Error :("
				info_for_dialog_backtitle=$info_for_dialog_screen
			else
				echo ""
				echo -e "Login to server \e[41mError :(\e[0m"			
			fi
			
			# our JOSN response from server postion in array we extract that from headers and add to variable
			josn_final=${array3[40]}
			echo $josn_final  | jq
			
elif [[ $sucesfully_login_true_or_false =~ "true"  ]];
	then		
			if [ ! -z "$DIALOG" ];
			then
				info_for_dialog_screen="Login to server OK :)"
				info_for_dialog_backtitle=$info_for_dialog_screen
			else		
				echo ""
				echo -e "Login to server \e[42mOK :)\e[0m"
			fi
			
			# our JOSN response from server postion in array we extract that from headers and add to variable
			josn_final=${array3[39]}
			echo $josn_final  | jq
else

	if [ ! -z "$DIALOG" ];
	then
		info_for_dialog_screen="Problems with conection to Huawei Server" 
		info_for_dialog_backtitle=$info_for_dialog_screen
	else
		echo ""
		echo -e "Problems with conection to Huawei Server" 
	fi
fi
```

**Reply by xvanhien** [reply]
I think the line "if [[ $sucesfully_login_true_or_false == "true"  ]];" is an issue. It also not working:
Please wait connecting!
Connection and login to API with given Username&Password
sucesfully_login_true_or_false:
true
value{array3[39] =
value{array3[40] = {"data":null,"success":true,"failCode":0,"params":{},"message":null}
value{array3[@] = HTTP/1.1 200 OK Server: product only Date: Fri, 28 May 2021 06:42:14 GMT Content-Type: application/json;charset=UTF-8 Transfer-Encoding: chunked Connection: keep-alive set-cookie: XSRF-TOKEN=x-vymnmnbutgqrc8albubt07petiuqpeqk9ihc8b7t6onyjv6moarvdiaq9jum4aeqbzvvtgg4s8g86rannxamns5dvxg4vx9fns9fhfbvml05ikpeo9ldk47soaqm2qg8;Path=/;Secure x-trace-enable: false x-xss-protection: 1; mode=block x-content-type-options: nosniff xsrf-token: x-vymnmnbutgqrc8albubt07petiuqpeqk9ihc8b7t6onyjv6moarvdiaq9jum4aeqbzvvtgg4s8g86rannxamns5dvxg4vx9fns9fhfbvml05ikpeo9ldk47soaqm2qg8 x-frame-options: SAMEORIGIN x-download-options: noopen strict-transport-security: max-age=31536000; includeSubDomains content-encoding: gzip  {"data":null,"success":true,"failCode":0,"params":{},"message":null}

Problems with conection to Huawei Server
showing response JOSN from login API:

Network Error :(


When I make a comment to this, it works:
if [[ $sucesfully_login_true_or_false == "true"  ]];
then
	local sucesfully_login_true_or_false=$(echo ''${array3[39]}''  | jq '.success' )
else
	local sucesfully_login_true_or_false=$(echo ''${array3[40]}''  | jq '.success' )
	
fi

**Reply by BlazejosP** [reply]
I'm during solving this problem. As for now I'm rewriting this part of login function to be independent from position of certain word in answer array. In one two day I will upload that and hope that this solve problems like your definitely.
